GeometryCore/src/LinearAlgebra.cpp

```diff
@@ -4,29 +4,114 @@
+        intersection.y = (get<2>(stline1) * get<0>(stline2) - get<2>(stline2) * get<0>(stline1)) / (get<1>(stline2) * get<0>(stline1) - get<1>(stline1) * get<0>(stline2));
+        return intersection;
+    }
+
     // Return true if line1 is colliding with line2
-    bool lineSegmentsCollide(const LineSegment2D& line1, const LineSegment2D& line2)
-    {
-        return {};
+    bool lineSegmentsCollide(const LineSegment2D& line1, const LineSegment2D& line2) {
+        auto intersection = lineSegmentsIntersection(line1, line2);
+        if ((intersection.x >= std::min(line1.a.x, line1.b.x)) && (intersection.x <= std::max(line1.a.x, line1.b.x)) &&
+            (intersection.x >= std::min(line2.a.x, line2.b.x)) && (intersection.x <= std::max(line2.a.x, line2.b.x)) &&
+            (intersection.y >= std::min(line1.a.y, line1.b.y)) && (intersection.y <= std::max(line1.a.y, line1.b.y)) &&
+            (intersection.y >= std::min(line2.a.y, line2.b.y)) && (intersection.y <= std::max(line2.a.y, line2.b.y)))
+        {
+            return true;
+        }
+        return false;
     }
 
     // Return distance from the p to the line
-    FPoint2D::coord distanceFromPointToLine(const FPoint2D& p, const LineSegment2D& line)
-    {
-        return {};
+    FPoint2D::coord distanceFromPointToLine(const FPoint2D& p, const LineSegment2D& line) {
+        auto projection = pointProjectionToLineSegment(p, line);
+
+        if ((projection.x >= std::min(line.a.x, line.b.x)) && (projection.x <= std::max(line.a.x, line.b.x)) &&
+            (projection.y >= std::min(line.a.y, line.b.y)) && (projection.y <= std::max(line.a.y, line.b.y))) {
+            return distance(projection, p);
+        }
+        else {
+            return std::min(distance(line.a, p), distance(line.b, p));
+        }
+    }
+
+    FPoint2D::coord distanceBetweenLineSegments(const LineSegment2D& line1, const LineSegment2D& line2) {
+        auto check_projection = [](const FPoint2D& p, const LineSegment2D& line, FPoint2D::coord& minimal_distance) { minimal_distance = std::min(distanceFromPointToLine(p, line), minimal_distance); };
+
+        FPoint2D::coord min_dis = std::min(std::min(distance(line1.a, line2.a), distance(line1.a, line2.b)), std::min(distance(line1.b, line2.a), distance(line1.b, line2.b)));
+        check_projection(line2.a, line1, min_dis);
+        check_projection(line2.b, line1, min_dis);
+        check_projection(line1.a, line2, min_dis);
+        check_projection(line1.b, line2, min_dis);
+
+        return min_dis;
     }
 
     // Return true if poly and line are colliding with each other.
     // Note: consider collision even if one of line end lies exactly on the polygon contour
-    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line)
-    {
-        return {};
+    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line) {
+        auto convex = poly.convexHull();
```

Comment:
Why do we calculate convex hull for poly? What will be the result of this function in such case?
![image](https://user-images.githubusercontent.com/11663898/81966227-e0edf600-9621-11ea-9ec3-9a0684813ca5.png)
In fact, line doesn't collide with poly, but it collides with convex hull of poly. Will function return true for that case? (whereas it should return false)

Reply:
Oh, I dont think about this before, I will implement this